### PR TITLE
[RFC] Association name setter

### PR DESCRIPTION
**Pull request by robertpustulka**
Following [comments](https://github.com/cakephp/cakephp/pull/10142).

Association names are mutable, but that change isn't reflected in a collection nor the target table. Changing association name without manually reattaching it in a collection makes it unusable. `EagerLoader` [checks if these names match](https://github.com/cakephp/cakephp/blob/3.next/src/ORM/EagerLoader.php#L496) and throws an exception. Is this exception enough or should we disallow changing association name as it would probably break it?

In other words `AssociationCollection` key, `Association` name and target `Table` alias should be identical. Changing one of them would probably break things.

My way to improve this is to check if a new association name matches the target table alias **after** it has been resolved.

There's also an issue with `AssociationCollection` not reflecting association name change, but I'm not sure we can do anything about it without changing the way associations are collected.

### CODE REVIEW

**Review by markstory — CHANGES_REQUESTED** [review]
Looks ok to me, but needs some tests.

**Comment by robertpustulka** [comment]
Tests added.

**Comment by lorenzo** [comment]
thanks @robertpustulka

**Comment by inoas** [comment]
Love these kind of changes. Fail early makes my heart feel all fuzzy and secure!
